cab-token-generator/javatests/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactoryTest.java

```diff
@@ -0,0 +1,954 @@
+    triggerConcurrentRefresh(factory, 3);
+
+    // After all threads complete, the request count should be 2 (the initial fetch plus one
+    // blocking refresh).
+    assertEquals(2, mockStsTransportFactory.transport.getRequestCount());
+  }
+
+  @Test
+  public void refreshCredentialsIfRequired_asyncMultiThread()
+      throws IOException, InterruptedException {
+    final ClientSideCredentialAccessBoundaryFactory factory =
+        getClientSideCredentialAccessBoundaryFactory(RefreshType.ASYNC);
+
+    // Call refreshCredentialsIfRequired() once to initialize the intermediate credentials. This
+    // should make one request to the STS endpoint.
+    factory.refreshCredentialsIfRequired();
+    assertEquals(1, mockStsTransportFactory.transport.getRequestCount());
+
+    // Simulate multiple threads concurrently calling refreshCredentialsIfRequired(). In async
+    // mode, the first call should trigger a background refresh. Subsequent calls should NOT
+    // trigger additional refreshes while the background refresh is still pending.
+    triggerConcurrentRefresh(factory, 5);
+
+    // Introduce a small delay to allow the asynchronous refresh task to complete.
+    try {
+      Thread.sleep(100);
+    } catch (InterruptedException e) {
+      throw new IOException(e);
+    }
+
+    // After the delay, the request count should be 2, indicating that the initial fetch and a
+    // single async refresh occurred (not one per thread).
+    assertEquals(2, mockStsTransportFactory.transport.getRequestCount());
+  }
+
+  @Test
+  public void refreshCredentialsIfRequired_sourceCredentialCannotRefresh_throwsIOException()
+      throws Exception {
+    // Simulate error when refreshing the source credential.
+    mockTokenServerTransportFactory.transport.setError(new IOException());
+
+    GoogleCredentials sourceCredentials =
+        getServiceAccountSourceCredentials(mockTokenServerTransportFactory);
+
+    ClientSideCredentialAccessBoundaryFactory factory =
+        ClientSideCredentialAccessBoundaryFactory.newBuilder()
+            .setSourceCredential(sourceCredentials)
+            .setHttpTransportFactory(mockStsTransportFactory)
+            .build();
+
+    IOException exception = assertThrows(IOException.class, factory::refreshCredentialsIfRequired);
+    assertEquals("Unable to refresh the provided source credential.", exception.getMessage());
+  }
+
+  // Tests related to the builder methods
+  @Test
+  public void builder_noSourceCredential_throws() {
+    NullPointerException exception =
+        assertThrows(
+            NullPointerException.class,
+            () ->
+                ClientSideCredentialAccessBoundaryFactory.newBuilder()
+                    .setHttpTransportFactory(OAuth2Utils.HTTP_TRANSPORT_FACTORY)
+                    .build());
+    assertEquals("Source credential must not be null.", exception.getMessage());
+  }
+
+  @Test
+  public void builder_minimumTokenLifetime_negative_throws() throws IOException {
```

Comment:
Add tests for asserting minimum token lifetime behavior & refresh margin behavior (I might've missed it so resolve if so)